src/storage-client/src/controller.rs

```diff
@@ -2676,56 +2691,65 @@ where
+        let finalized_shards: BTreeSet<ShardId> = futures::stream::iter(shards)
             .map(|shard_id| async move {
-                let (mut write, mut critical_since_handle) = this
-                    .open_data_handles(
-                        "finalizing shards",
+                // Open read handle, whose since is the global since.
+                let read_handle: ReadHandle<SourceData, (), T, Diff> = persist_client
+                    .open_leased_reader(
                         shard_id,
-                        RelationDesc::empty(),
-                        persist_client,
+                        "finalizing shards",
+                        Arc::new(RelationDesc::empty()),
+                        Arc::new(UnitSchema),
                     )
-                    .await;
-
-                let our_epoch = PersistEpoch::from(this.state.envd_epoch);
-
-                match critical_since_handle
-                    .compare_and_downgrade_since(&our_epoch, (&our_epoch, &Antichain::new()))
                     .await
-                {
-                    Ok(_) => info!("successfully finalized read handle for shard {shard_id:?}"),
-                    Err(e) => mz_ore::halt!("fenced by envd @ {e:?}. ours = {our_epoch:?}"),
-                }
-
-                if write.upper().is_empty() {
-                    info!("write handle for shard {:?} already finalized", shard_id);
-                } else {
-                    write
-                        .append(
-                            Vec::<((crate::types::sources::SourceData, ()), T, Diff)>::new(),
-                            write.upper().clone(),
-                            Antichain::new(),
+                    .expect("invalid persist usage");
+
+                // If global since is empty, we can close shard because no one has an outstanding
+                // read hold.
+                if read_handle.since().is_empty() {
+                    let mut write_handle: WriteHandle<SourceData, (), T, Diff> = persist_client
+                        .open_writer(
+                            shard_id,
+                            "finalizing shards",
+                            Arc::new(RelationDesc::empty()),
+                            Arc::new(UnitSchema),
                         )
                         .await
-                        .expect("failed to connect")
-                        .expect("failed to truncate write handle");
+                        .expect("invalid persist usage");
+
+                    if !write_handle.upper().is_empty() {
+                        write_handle
+                            .append(
+                                Vec::<((crate::types::sources::SourceData, ()), T, Diff)>::new(),
+                                write_handle.upper().clone(),
+                                Antichain::new(),
+                            )
+                            .await
+                            .expect("failed to connect")
```

Comment:
This could fail and we could handle it gracefully by just retrying next time `finalize_shards` is called, meaning we'd just not panic here because the rest already works?